### PR TITLE
[IMP] hr_holidays: improve extendability

### PR DESCRIPTION
**Pull request by davidtranhp**
Split the method `_create_resource_leave()` that does value preparation and record creation at the same time into 2 methods:

- [x] `_prepare_resource_leave_vals_list()` that will prepare list of value for creation later in `_create_resource_leave()`
- [x] `_create_resource_leave()` now calls the `_prepare_resource_leave_vals_list()` instead of having done the job itself

This structure give convenient for others to extend Odoo functionality without harming performance and less attempt

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by davidtranhp** [comment]
@mart-e can you please take a look into this? Is this possible for v13 as it has no side-effect or should it target master only?

**Comment by mart-e** [comment]
@davidtranhp can't you work with the existing method? It's already a small method that is easy to override.
I let the application team decide what to do with this kind of PR. In this case it's @tivisse 's team

**Comment by davidtranhp** [comment]
> @davidtranhp can't you work with the existing method? It's already a small method that is easy to override. I let the application team decide what to do with this kind of PR. In this case it's @tivisse 's team

I can, but perfomance would be better if it is allowed to inject/manipulate data before writing to db.

**Comment by davidtranhp** [comment]
> Hello @davidtranhp
> 
> I'm not against the idea, but please keep the method working in batch (to avoid breaking the prefetch), aka no self.ensure_one().
> 
> After that, I'm ok to r+ it.
> 
> Have a nice day,
> 
> Yannick.

I've made it. `self.ensure_one()` was removed. Thanks @tivisse

File: addons/hr_holidays/models/hr_leave.py [hunk 1]
```diff
@@ -791,19 +791,24 @@ def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=Fals
...
+        """
+        return {
+            'name': self.name,
+            'date_from': self.date_from,
+            'holiday_id': self.id,
+            'date_to': self.date_to,
+            'resource_id': self.employee_id.resource_id.id,
+            'calendar_id': self.employee_id.resource_calendar_id.id,
+            'time_type': self.holiday_status_id.time_type,
+            }
```
Comment (tivisse) [inline addons/hr_holidays/models/hr_leave.py:805]: I meant 
```suggestion
        return [{
            'name': leave.name,
            'date_from': leave.date_from,
            'holiday_id': leave.id,
            'date_to': leave.date_to,
            'resource_id': leave.employee_id.resource_id.id,
            'calendar_id': leave.employee_id.resource_calendar_id.id,
            'time_type': leave.holiday_status_id.time_type,
            } for leave in self]
```

Comment (davidtranhp) [reply addons/hr_holidays/models/hr_leave.py]: I made it. Tks!

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**